File: ALittle/Module/ALittleIDE/Other/GameLibrary/Std/JSScript/Std/Net/HttpFileSenderTemplate.js
{
if (typeof ALittle === "undefined") ALittle = {};


ALittle.IHttpFileSenderNative = JavaScript.Class(undefined, {
	GetID : function() {
		return 0;
	},
	SetURL : function(url, file_path, download, start_size) {
	},
	Start : function() {
	},
	Stop : function() {
	},
	GetPath : function() {
		return "";
	},
}, "ALittle.IHttpFileSenderNative");

let __HttpFileSenderMap = new Map();
if (ALittle.IHttpFileSender === undefined) throw new Error(" extends class:ALittle.IHttpFileSender is undefined");
ALittle.HttpFileSenderTemplate = JavaScript.Class(ALittle.IHttpFileSender, {
	Ctor : function(ip, port, file_path, start_size, callback) {
		this._interface = ALittle.NewObject(this.__class.__element[0]);
		this._ip = ip;
		this._port = port;
		if (this._port === undefined) {
			this._port = 80;
		}
		this._file_path = file_path;
		this._start_size = start_size;
		this._callback = callback;
		this._cur_size = 0;
		this._total_size = 0;
	},
	SendDownloadRPC : function(thread, method, content) {
		this._thread = thread;
		__HttpFileSenderMap.set(this._interface.GetID(), this);
		if (this._start_size === undefined) {
			this._start_size = 0;
		}
		let url = "http://" + this._ip + ":" + this._port + "/" + method;
		this._interface.SetURL(this.HttpUrlAppendParamMap(url, content), this._file_path, true, this._start_size);
		this._interface.Start();
	},
	SendUploadRPC : function(thread, method, content) {
		this._thread = thread;
		__HttpFileSenderMap.set(this._interface.GetID(), this);
		if (this._start_size === undefined) {
			this._start_size = 0;
		}
		let url = "http://" + this._ip + ":" + this._port + "/" + method;
		this._interface.SetURL(this.HttpUrlAppendParamMap(url, content), this._file_path, false, this._start_size);
		this._interface.Start();
	},
	Stop : function() {
		this._interface.Stop();
	},
	GetTotalSize : function() {
		return this._total_size;
	},
	GetCurSize : function() {
		return this._cur_size;
	},
	HandleSucceed : function() {
		__HttpFileSenderMap.delete(this._interface.GetID());
		let [result, reason] = ALittle.Coroutine.Resume(this._thread, undefined, this);
		if (result !== true) {
			ALittle.Error(reason);
		}
	},
	HandleFailed : function(reason) {
		__HttpFileSenderMap.delete(this._interface.GetID());
		let [result, error] = ALittle.Coroutine.Resume(this._thread, reason, this);
		if (result !== true) {
			ALittle.Error(error);
		}
	},
	HandleProcess : function(cur_size, total_size) {
		this._cur_size = cur_size;
		this._total_size = total_size;
		if (this._callback !== undefined) {
			this._callback(this._interface);
		}
	},
	HttpUrlAppendParamMap : function(url, param) {
		let list = [];
		let count = 0;
		let ___OBJECT_1 = param;
		for (let key in ___OBJECT_1) {
			let value = ___OBJECT_1[key];
			if (value === undefined) continue;
			++ count;
			list[count - 1] = key + "=" + ALittle.String_ToString(value);
		}
		if (ALittle.String_Find(url, "?") === undefined) {
			url = url + "?";
		} else {
			url = url + "&";
		}
		return url + ALittle.String_Join(list, "&");
	},
}, "ALittle.HttpFileSenderTemplate");

ALittle.__ALITTLEAPI_HttpFileSucceed = function(id) {
	let client = __HttpFileSenderMap.get(id);
	if (client === undefined) {
		return;
	}
	client.HandleSucceed();
}

ALittle.__ALITTLEAPI_HttpFileFailed = function(id, reason) {
	let client = __HttpFileSenderMap.get(id);
	if (client === undefined) {
		return;
	}
	client.HandleFailed(reason);
}

ALittle.__ALITTLEAPI_HttpFileProcess = function(id, cur_size, total_size) {
	let client = __HttpFileSenderMap.get(id);
	if (client === undefined) {
		return;
	}
	client.HandleProcess(cur_size, total_size);
}

}

File: ALittle/Module/Std/Script/Std/Net/HttpFileSenderTemplate.lua
-- ALittle Generate Lua
module("ALittle", package.seeall)

local ___thispath = select('1', ...):match("(.+[/\\]).+$") or ""
local ___rawset = rawset
local ___pairs = pairs
local ___ipairs = ipairs


IHttpFileSenderNative = Lua.Class(nil, "ALittle.IHttpFileSenderNative")

function IHttpFileSenderNative:GetID()
	return 0
end

function IHttpFileSenderNative:SetURL(url, file_path, download, start_size)
end

function IHttpFileSenderNative:Start()
end

function IHttpFileSenderNative:Stop()
end

function IHttpFileSenderNative:GetPath()
	return ""
end

local __HttpFileSenderMap = {}
assert(ALittle.IHttpFileSender, " extends class:ALittle.IHttpFileSender is nil")
HttpFileSenderTemplate = Lua.Class(ALittle.IHttpFileSender, "ALittle.HttpFileSenderTemplate")

function HttpFileSenderTemplate:Ctor(ip, port, file_path, start_size, callback)
	___rawset(self, "_interface", self.__class.__element[1]())
	___rawset(self, "_ip", ip)
	___rawset(self, "_port", port)
	if self._port == nil then
		___rawset(self, "_port", 80)
	end
	___rawset(self, "_file_path", file_path)
	___rawset(self, "_start_size", start_size)
	___rawset(self, "_callback", callback)
	___rawset(self, "_cur_size", 0)
	___rawset(self, "_total_size", 0)
end

function HttpFileSenderTemplate:SendDownloadRPC(thread, method, content)
	self._thread = thread
	__HttpFileSenderMap[self._interface:GetID()] = self
	if self._start_size == nil then
		self._start_size = 0
	end
	local url = self._ip .. ":" .. self._port .. "/" .. method
	if String_Find(self._ip, "http://") ~= 1 and String_Find(self._ip, "https://") ~= 1 then
		if self._port == 443 then
			url = "https://" .. url
		else
			url = "http://" .. url
		end
	end
	self._interface:SetURL(self:HttpUrlAppendParamMap(url, content), self._file_path, true, self._start_size)
	self._interface:Start()
end

function HttpFileSenderTemplate:SendUploadRPC(thread, method, content)
	self._thread = thread
	__HttpFileSenderMap[self._interface:GetID()] = self
	if self._start_size == nil then
		self._start_size = 0
	end
	local url = "http://" .. self._ip .. ":" .. self._port .. "/" .. method
	self._interface:SetURL(self:HttpUrlAppendParamMap(url, content), self._file_path, false, self._start_size)
	self._interface:Start()
end

function HttpFileSenderTemplate:Stop()
	self._interface:Stop()
end

function HttpFileSenderTemplate:GetFilePath()
	return self._file_path
end

function HttpFileSenderTemplate:GetTotalSize()
	return self._total_size
end

function HttpFileSenderTemplate:GetCurSize()
	return self._cur_size
end

function HttpFileSenderTemplate:HandleSucceed()
	__HttpFileSenderMap[self._interface:GetID()] = nil
	local result, reason = Coroutine.Resume(self._thread, nil)
	if result ~= true then
		Error(reason)
	end
end

function HttpFileSenderTemplate:HandleFailed(reason)
	__HttpFileSenderMap[self._interface:GetID()] = nil
	local result, error = Coroutine.Resume(self._thread, reason)
	if result ~= true then
		Error(error)
	end
end

function HttpFileSenderTemplate:HandleProcess(cur_size, total_size)
	self._cur_size = cur_size
	self._total_size = total_size
	if self._callback ~= nil then
		self._callback(self)
	end
end

function HttpFileSenderTemplate:HttpUrlAppendParamMap(url, param)
	if param == nil then
		return url
	end
	local list = {}
	local count = 0
	for key, value in ___pairs(param) do
		count = count + 1
		list[count] = key .. "=" .. String_ToString(value)
	end
	if String_Find(url, "?") == nil then
		url = url .. "?"
	else
		url = url .. "&"
	end
	return url .. String_Join(list, "&")
end

function __ALITTLEAPI_HttpFileSucceed(id)
	local client = __HttpFileSenderMap[id]
	if client == nil then
		return
	end
	client:HandleSucceed()
end

function __ALITTLEAPI_HttpFileFailed(id, reason)
	local client = __HttpFileSenderMap[id]
	if client == nil then
		return
	end
	client:HandleFailed(reason)
end

function __ALITTLEAPI_HttpFileProcess(id, cur_size, total_size)
	local client = __HttpFileSenderMap[id]
	if client == nil then
		return
	end
	client:HandleProcess(cur_size, total_size)
end

function DownloadFile(ip, port, method, file_path)
local ___COROUTINE = coroutine.running()
	local sender
	sender = Lua.Template(HttpFileSenderTemplate, "ALittle.HttpFileSenderTemplate<Lua.LuaHttpFileInterface>", Lua.LuaHttpFileInterface)(ip, port, file_path, 0)
	return IHttpFileSender.InvokeDownload(method, sender, nil)
end

function UploadFile(ip, port, method, file_path)
local ___COROUTINE = coroutine.running()
	local sender
	sender = Lua.Template(HttpFileSenderTemplate, "ALittle.HttpFileSenderTemplate<Lua.LuaHttpFileInterface>", Lua.LuaHttpFileInterface)(ip, port, file_path, 0)
	local error = IHttpFileSender.InvokeUpload(method, sender, nil)
	return error
end

